File: LeetCode/Easy/0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int maxDiameter = 0;
    public int diameterOfBinaryTree(TreeNode root) {
        traverse(root);
        return maxDiameter;
        
    }
    void traverse(TreeNode root ){
        if(root == null){
            return;
        }
        int leftDepthMax = depthMax(root.left);
        int rightDepthMax = depthMax(root.right);
        int curDiameter = leftDepthMax + rightDepthMax;
        maxDiameter = Math.max(curDiameter, maxDiameter);
        traverse(root.left);
        traverse(root.right);


    }
    int depthMax(TreeNode root){
        if(root == null){
            return 0;
        }
        int leftMax = depthMax(root.left);
        int rightMax = depthMax(root.right);
        int depthMax = Math.max(leftMax, rightMax) + 1;
        return depthMax;
    }
}

File: LeetCode/Easy/0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.maxDiameter = 0
        self.maxDepth(root)

        return self.maxDiameter

    def maxDepth(self, root):
        if root is None:
            return 0
        leftMax = self.maxDepth(root.left)
        rightMax = self.maxDepth(root.right)
        # 只有在 maxDepth 里，才同时拥有 leftMax 和 rightMax
        currentDiameter = leftMax + rightMax
        self.maxDiameter = max(self.maxDiameter, currentDiameter)

        return 1 + max(leftMax, rightMax)

        

        